flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+			this.webOverviewDir = new Path(webDir.getAbsolutePath(), OVERVIEWS_FOLDER_NAME);
+		}
+
+		@Override
+		public void run() {
+			for (Iterator<Path> iterator = refreshDirs.iterator(); iterator.hasNext();) {
+				Path refreshDir = iterator.next();
+				FileSystem fs;
+				try {
+					fs = refreshDir.getFileSystem();
+				} catch (IOException e) {
+					LOG.error("Failed to create FileSystem for path {}.", refreshDir, e);
+					// there's most likely something wrong with the path itself
+					try {
+						iterator.remove();
+					} catch (Exception ignored) {
+						//well, we tried
+					}
+					continue;
+				}
+
+				/**
+				 * The expected directory structure is as follows:
+				 * /:refreshDir/:jobid/jobs/:jobid/...
+				 * /:refreshDir/:jobid/jobs/:jobid.json
+				 * /:refreshDir/:jobid/joboverview.json
+				 */
+				// contents of /:refreshDir
+				FileStatus[] jobArchives;
+				try {
+					jobArchives = fs.listStatus(refreshDir);
+				} catch (IOException e) {
+					LOG.error("Failed to access job archive location for path {}.", refreshDir, e);
+					continue;
+				}
+				boolean updateOverview = false;
+				for (FileStatus jobArchive : jobArchives) {
+					Path jobArchivePath = jobArchive.getPath();
+					String jobID = jobArchivePath.getName();
+					if (cachedArchives.put(jobID, refreshDir) == null) {
+						try {
+							// contents of /:refreshDir/:jobid
+							FileStatus[] archiveFiles = fs.listStatus(jobArchivePath);
+							for (FileStatus archiveFile : archiveFiles) { 
+								if (archiveFile.isDir()) {
+									// contents of /:refreshDir/:jobid/jobs
+									for (FileStatus jobFile : fs.listStatus(archiveFile.getPath())) {
+										/**
+										 * src: /:refreshDir/:jobid/jobs/X
+										 * dst: /:webDir/jobs/X
+										 * 
+										 * X is either :jobid.json or the :jobid directory.
+										 */
+										FileCache.copy(jobFile.getPath(), new Path(webJobDir + "/" + jobFile.getPath().getName()), false);
+									}
+								} else {
+									/**
+									 * src: /:refreshDir/:jobid/joboverview.json
+									 * dst: /:webDir/joboverviews/:jobid.json
+									 */
+									FileCache.copy(archiveFile.getPath(), new Path(webOverviewDir, jobID + ".json"), false);
+									updateOverview = true;
+								}
+							}
+						} catch (IOException e) {
+							LOG.error("Failed to fetch archive file.", e);
+							cachedArchives.remove(jobID);
+							continue;
+						}
+					}
+				}
+				if (updateOverview) {
+					updateJobOverview(webDir);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Backup file fetcher that is used by the {@link HistoryServerStaticFileServerHandler} if it cannot find a file
+	 * within a job archive. This class does not share code with the {@link JobArchiveFetcherTask} since the former
+	 * fetches entire archives at a given location based on a path whereas this class only fetches one specific file
+	 * based on a REST URL.
+	 */
+	public static class JobFileFetcher {
```

Comment:
This is only relevant for the history server static file handler. I would move it there.